virt_lightning/api.py

```diff
@@ -0,0 +1,442 @@
+
+        if domain.context != context:
+            continue
+
+        template = ssh_host_template
+
+        output += template.format(
+            name=domain.name,
+            username=domain.username,
+            ipv4=domain.ipv4.ip,
+            ssh_key_file=domain.ssh_key,
+        )
+
+    for group_name, domains in groups.items():
+        output += "\n[{group_name}]".format(group_name=group_name)
+        for domain in domains:
+            output += domain.name
+    return output
+
+
+def status(configuration, context="default", name=None, **kwargs):
+    """
+    Returns the status of the VM of the envionment
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+
+    for domain in hv.list_domains():
+        if context and context != domain.context:
+            continue
+        name = domain.name
+        if not domain.ipv4:  # Not a VL managed VM
+            continue
+        yield {
+            "name": name,
+            "ipv4": domain.ipv4 and str(domain.ipv4.ip),
+            "context": domain.context,
+            "username": domain.username,
+        }
+
+
+def exec_ssh(configuration, name=None, **kwargs):
+    """
+    Open an SSH connection on a host
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.get_domain_by_name(name).exec_ssh()
+
+
+def list_domains(configuration, name=None, **kwargs):
+    """
+    Return a list Python-libvirt instance of the running libvirt VM.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    return sorted(hv.list_domains())
+
+
+def down(configuration, context="default", **kwargs):
+    """
+    Stop and remove a running environment.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_network(configuration.network_name, configuration.network_cidr)
+    hv.init_storage_pool(configuration.storage_pool)
+    for domain in hv.list_domains():
+        if domain.context != context:
+            continue
+        logger.info("%s purging %s", symbols.TRASHBIN.value, domain.name)
+        hv.clean_up(domain)
+
+    if bool(distutils.util.strtobool(configuration.network_auto_clean_up)):
+        hv.network_obj.destroy()
+
+
+def distro_list(configuration, **kwargs):
+    """
+    Return a list of VM images that are available on the system.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_storage_pool(configuration.storage_pool)
+    return hv.distro_available()
+
+
+def storage_dir(configuration, **kwargs):
+    """
+    Return the location of the VM image storage directory.
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_storage_pool(configuration.storage_pool)
+    return hv.get_storage_dir()
+
+
+def fetch(configuration, progress_callback=None, **kwargs):
```

Comment:
If code call this function from async code, I recommend rewrite that to use aiohttp because urllib is blocking library. Depends on case what you want to reach.

Reply:
I agree, but `aiohttp` is also a new dependency. I would prefer to avoid that by default.